# Trump Tweets

This the data behind the story [The World’s Favorite Donald Trump Tweets](https://fivethirtyeight.com/features/the-worlds-favorite-donald-trump-tweets/).

In [1]:
# The usual suspects ...
import logging
import pandas as pd

# And their accomplices ...
from gensim import corpora
from gensim import models
from gensim import similarities
from nltk.corpus import stopwords
from nltk.tokenize import wordpunct_tokenize
from collections import defaultdict
from pprint import pprint

# Settings
logging.basicConfig(format='%(asctime)s : %(levelname)s : %(message)s', level=logging.INFO)

In [2]:
tweets = pd.read_csv('realDonaldTrump_poll_tweets.csv')

In [3]:
tweets.shape

(448, 3)

In [4]:
tweets.head()

id          created_at  \
0  7.656299e+17  8/16/2016 19:22:57   
1  7.587319e+17  7/28/2016 18:32:31   
2  7.583505e+17  7/27/2016 17:16:56   
3  7.575775e+17  7/25/2016 14:05:27   
4  7.536034e+17  7/14/2016 14:53:46   

                                                text  
0  It's just a 2-point race, Clinton 38%, Trump 3...  
1  "@LallyRay: Poll: Donald Trump Sees 17-Point P...  
2  Great new poll - thank you!\n#MakeAmericaGreat...  
3  Great POLL numbers are coming out all over. Pe...  
4  Another new poll. Thank you for your support! ...

In [5]:
# Text corpus
document = [i for i in tweets['text']]

In [6]:
# Removing common words and tokenize
stop_words = set(stopwords.words('english'))
stop_words.update(['-', '=', '+', '*','.', ',', '"', "'", '?', '!', ':', ';', '(', ')', '[', ']', '{', '}'])
for doc in document:
    list_of_words = [i.lower() for i in wordpunct_tokenize(doc) if i.lower() not in stop_words]
stop_words.update(list_of_words)

In [7]:
# Removing common words
texts = [[word for word in doc.lower().split() if word not in stop_words] for doc in document]

# Removing words that appear only once
frequency = defaultdict(int)
for text in texts:
    for token in text:
        frequency[token] += 1

texts = [[token for token in text if frequency[token] > 1] for text in texts]

# Printing the top 10
pprint(texts[:10])

[['clinton', 'trump'],
 ['poll:',
  'donald',
  'trump',
  'two',
  'breitbart',
  '@realdonaldtrump"',
  'great!'],
 ['great', 'new', 'poll', 'thank', 'you!', '#makeamericagreatagain'],
 ['great',
  'poll',
  'numbers',
  'coming',
  'people',
  'want',
  'another',
  'four',
  'years',
  'crooked',
  'hillary',
  'even'],
 ['another', 'new', 'poll.', 'thank', 'support!', '#imwithyou'],
 ['great', 'new', 'poll-', 'thank', 'america!', '#trump2016', '#imwithyou'],
 ['despite', 'spending', 'day', 'ads', 'nationwide', 'zero'],
 ['great', 'poll-', 'thank', 'you!'],
 ['new', 'poll', 'thank', 'you!', '#trump2016'],
 ['new',
  'q',
  'poll',
  'going',
  'win',
  'make',
  'america',
  'great',
  'again!',
  '#trump2016']]


In [8]:
# Create dictionary of document
bag = corpora.Dictionary(texts)
bag.save('trump.dict')

# Converting document to a vector (bag-of-words)
corpus = [bag.doc2bow(text) for text in texts]
corpora.MmCorpus.serialize('trump.mm', corpus)

2019-02-16 20:36:34,257 : INFO : adding document #0 to Dictionary(0 unique tokens: [])
2019-02-16 20:36:34,267 : INFO : built Dictionary(660 unique tokens: ['clinton', 'trump', '@realdonaldtrump"', 'breitbart', 'donald']...) from 448 documents (total 3742 corpus positions)
2019-02-16 20:36:34,267 : INFO : saving Dictionary object under trump.dict, separately None
2019-02-16 20:36:34,269 : INFO : saved trump.dict
2019-02-16 20:36:34,275 : INFO : storing corpus in Matrix Market format to trump.mm
2019-02-16 20:36:34,277 : INFO : saving sparse matrix to trump.mm
2019-02-16 20:36:34,278 : INFO : PROGRESS: saving document #0
2019-02-16 20:36:34,291 : INFO : saved 448x660 matrix, density=1.239% (3664/295680)
2019-02-16 20:36:34,293 : INFO : saving MmCorpus index to trump.mm.index


We have assigned a unique integer id to all words appearing in the corpus by:
   
   1. sweeping across the texts
   2. collecting word counts and relevant statistics
   
Our corpus is a 448 x 661 matrix.

***

### Transformation: _tf-idf_

#### Step 1:

In [9]:
# Initialization
tfidf = models.TfidfModel(corpus)

2019-02-16 20:36:40,813 : INFO : collecting document frequencies
2019-02-16 20:36:40,816 : INFO : PROGRESS: processing document #0
2019-02-16 20:36:40,820 : INFO : calculating IDF weights for 448 documents and 659 features (3664 matrix non-zeros)


We have initialized (trained) a transaformation model. Different transformation may require different initialization parameters; however, in our case, ___tf-idf___, the "training" consists simply of going through the supplied corpus once and computing document frequencies of all its features. This is in comparison to ___Latent Semantic Analysis___ & ___Latent Dirichlet Allocation___ which are more involved and take more time.

|Note:|
|---|
|**A note on transaformations**<br>Transformations always convert between two specific vector spaces. The same vector space (= the same set of feature ids) must be used for training as well as for subsequent vector transformations. Failure to use the same input feature space, such as applying a different string preprocessing, using different feature ids, or using bag-of-words input vectors where ___tf-idf___ vectors are expceted, will result in feature mismatch during transformation calls and consequently in either garbage output and/or runtime exceptions.|

#### Step 2:
From now on, ___tf-idf___ is treated as a read-only object that can be used to convert any vector from the old representation (___bag-of-words___ integer counts) to the new representation (___tf-idf___ real-valued weights).

In [10]:
# Applying the transformation to the whole corpus
corpus_tfidf = tfidf[corpus]

We have transformed our corpus (the one we used for training) into a weighted vector. We can do this for any vector (provided they come from the same vector space), even if they are not used in the corpus at all. This can be achived by _folding-in_ for ___LSA___ and by _topic inference_ for ___LDA___.

#### Step 3:
We will transform our ___tf-idf___ corpus via [Latent Semantic Indexing](https://en.wikipedia.org/wiki/Latent_semantic_indexing) into a latent 10-D space (... num_topics = 10).

In [11]:
# Initializing an LSI transformation
lsi = models.LsiModel(corpus_tfidf, id2word=bag, num_topics=10)
corpus_lsi = lsi[corpus_tfidf]

2019-02-16 20:36:44,774 : INFO : using serial LSI version on this node
2019-02-16 20:36:44,777 : INFO : updating model with new documents
2019-02-16 20:36:44,808 : INFO : preparing a new chunk of documents
2019-02-16 20:36:44,816 : INFO : using 100 extra samples and 2 power iterations
2019-02-16 20:36:44,817 : INFO : 1st phase: constructing (660, 110) action matrix
2019-02-16 20:36:44,844 : INFO : orthonormalizing (660, 110) action matrix
2019-02-16 20:36:44,971 : INFO : 2nd phase: running dense svd on (110, 448) matrix
2019-02-16 20:36:45,067 : INFO : computing the final decomposition
2019-02-16 20:36:45,068 : INFO : keeping 10 factors (discarding 76.499% of energy spectrum)
2019-02-16 20:36:45,084 : INFO : processed documents up to #448
2019-02-16 20:36:45,086 : INFO : topic #0(3.938): 0.411*"thank" + 0.407*"you!" + 0.364*"#makeamericagreatagain" + 0.346*"#trump2016" + 0.305*"great" + 0.288*"new" + 0.196*"poll" + 0.143*"poll-" + 0.113*"poll." + 0.111*"national"
2019-02-16 20:36:45,08

In [12]:
lsi.print_topics()

2019-02-16 20:36:48,853 : INFO : topic #0(3.938): 0.411*"thank" + 0.407*"you!" + 0.364*"#makeamericagreatagain" + 0.346*"#trump2016" + 0.305*"great" + 0.288*"new" + 0.196*"poll" + 0.143*"poll-" + 0.113*"poll." + 0.111*"national"
2019-02-16 20:36:48,856 : INFO : topic #1(2.986): 0.402*"trump" + -0.241*"you!" + 0.227*"carson" + 0.227*"rubio" + -0.209*"#makeamericagreatagain" + -0.204*"thank" + 0.191*"donald" + 0.184*"leads" + 0.182*"cruz" + 0.164*"@realdonaldtrump"
2019-02-16 20:36:48,858 : INFO : topic #2(2.459): -0.499*"great" + 0.278*"trump" + 0.268*"#makeamericagreatagain" + -0.242*"new" + -0.211*"america" + 0.193*"#trump2016" + -0.188*"make" + -0.180*"again!" + -0.169*"big" + 0.162*"rubio"
2019-02-16 20:36:48,859 : INFO : topic #3(2.292): 0.367*"national" + -0.316*"great" + -0.257*"rubio" + 0.226*"#trump2016" + 0.216*"#makeamericagreatagain" + 0.215*"lead" + -0.207*"carson" + -0.201*"poll-" + -0.191*"trump" + -0.184*"you!"
2019-02-16 20:36:48,861 : INFO : topic #4(2.171): -0.260*"na

[(0,
  '0.411*"thank" + 0.407*"you!" + 0.364*"#makeamericagreatagain" + 0.346*"#trump2016" + 0.305*"great" + 0.288*"new" + 0.196*"poll" + 0.143*"poll-" + 0.113*"poll." + 0.111*"national"'),
 (1,
  '0.402*"trump" + -0.241*"you!" + 0.227*"carson" + 0.227*"rubio" + -0.209*"#makeamericagreatagain" + -0.204*"thank" + 0.191*"donald" + 0.184*"leads" + 0.182*"cruz" + 0.164*"@realdonaldtrump"'),
 (2,
  '-0.499*"great" + 0.278*"trump" + 0.268*"#makeamericagreatagain" + -0.242*"new" + -0.211*"america" + 0.193*"#trump2016" + -0.188*"make" + -0.180*"again!" + -0.169*"big" + 0.162*"rubio"'),
 (3,
  '0.367*"national" + -0.316*"great" + -0.257*"rubio" + 0.226*"#trump2016" + 0.216*"#makeamericagreatagain" + 0.215*"lead" + -0.207*"carson" + -0.201*"poll-" + -0.191*"trump" + -0.184*"you!"'),
 (4,
  '-0.260*"national" + -0.252*"leads" + 0.240*"@realdonaldtrump" + -0.224*"donald" + -0.204*"great" + 0.199*"numbers" + 0.189*"@cnn" + 0.168*"you!" + -0.153*"america" + 0.152*"thank"'),
 (5,
  '0.323*"rubio" + 0

#### Topics

We've transformed our corpus to have 10 topics according to ___LSI___:
<br>
**Topic 1**
> _"thank", "you!", "#makeamericagreatagain", "#trump2016", "great", "new", "poll", "-", "poll-", "poll", "national"_

It appears that ___"thank"___, ___"you"___, the hash-tags ___"#makeamericagreatagain"___ & ___"#trump2016"___ are all related and contribute the most in the direction of the first topic. While ___"poll"___ (with its variants) and ___"national"___ contribute the least.

**Topic 2**
> _"trump", "you!", "carson", "rubio", "#makeamericagreatagain", "thank", "donald", "leads", "cruz", "@realdonaldtrump"_

In the second topic, we have the words ___"trump"___, ___"you!"___, ___"carson"___, ___"rubio"___ & the hash-tag ___"#makeamericagreatagain"___ contributing the most. This topic associates the current American president, [Donald Trump](https://en.wikipedia.org/wiki/Donald_Trump), with his Republican critics and opposition during the primaries, [Ben Carson](https://en.wikipedia.org/wiki/Ben_Carson) and [Marco Rubio](https://en.wikipedia.org/wiki/Marco_Rubio). While [Ted Cruz](https://en.wikipedia.org/wiki/Ted_Cruz) mentions contributed the least.

**Topic 3**
> _"great", "trump", "#makeamericagreatagain", "new", "america", "#trump2016", "make", "again!", "big", "rubio"_

Here we have the president's name being associated with "greatness", "bigness" and "America". This is likely in association with his compaign slogan ["Make America Great Again"](https://en.wikipedia.org/wiki/Make_America_Great_Again).

**Topic 4**
> _"national", "great", "rubio", "#trump2016", "#makeamericagreatagain", "lead", "poll-", "carson", "trump", "you!"_

Topic number 4 has a theme that contains "national", "great", "Marco Rubio", "leading", "poll", "Ben Carson", "Donald Trump" and the hash-tags ___"#trump2016"___ & ___"#makeamericagreatagain"___. This is likely the topic related to Donald Trump making comments about his Republican opposition and references to him leading in the national polls.

**Topic 5**
> _"national", "leads", "@realdonaldtrump", "donald", "great", "numbers", "@cnn", "you!", "america", "thank"_

In this topic we have [Donald Trump](https://en.wikipedia.org/wiki/Donald_Trump) using his Twitter handle [___"@realdonaldtrump"___](https://twitter.com/realDonaldTrump) and tweeting about how he is leading in the national polls. There is reference to [___"@cnn"___](https://twitter.com/cnn) and giving thanks to the people of America.

**Topic 6**
> _"rubio", "carson", "donald", "cruz", "clinton", "poll:", "leads", "national", "@realdonaldtrump", "bush"_

This topic contains tweets on topics where Donald Trump likely is comparing himself to his opposition, that is, [Marco Rubio](https://en.wikipedia.org/wiki/Marco_Rubio), [Ben Carson](https://en.wikipedia.org/wiki/Ben_Carson), [Ted Cruz](https://en.wikipedia.org/wiki/Ted_Cruz), [Hillary Clinton](https://en.wikipedia.org/wiki/Hillary_Clinton), and [Jeb Bush](https://en.wikipedia.org/wiki/Jeb_Bush), in the national polls.

**Topic 7**
> _"new", "@realdonaldtrump", "big", "great", "you!", "clinton", "hillary", "#1", "america", "make"_

Topic 7 has [Donald Trump](https://en.wikipedia.org/wiki/Donald_Trump) tweeting about his Democrat opponent [Hillary Clinton](https://en.wikipedia.org/wiki/Hillary_Clinton) with references of "#1", "new-ness", "big-ness", "great-ness", "making" and "America". This is likely tweets where Donald Trump is critisizing Hillary, then making reference to his campaign slogan.

**Topic 8**
> _"debate", "poll.", "said", "lead", "iowa", "new", "debate.", "@cnn", "@realdonaldtrump", "every"_

It appears in this topic, ___"debate"___ contributes the most. This is followed by lines (or sentences) ending with ___"poll"___, containing the words ___"said"___, ___"lead"___, ___"Iowa"___, ___"new"___, ___"every"___ and the CNN hash-tag ___"@cnn"___. [Iowa](https://en.wikipedia.org/wiki/Iowa) is referenced likely because of the presidential caucus, which is the first in the country and the starting point along with the [New Hampshire](https://en.wikipedia.org/wiki/New_Hampshire) primary, where the two major-party candidates for president are chosen.

**Topic 9**
> _"@realdonaldtrump", "debate", "big", "leads", "national", "lead", "last", "leading", "#1", "said"_

Topic 9 associates Donald Trump's tweets on the debate and leading in the national polls. The official handle [***@realdonaldtrump***](https://twitter.com/realDonaldTrump) contributes the most in this topic, likely because these tweets were made using this account.

**Topic 10**
> _"numbers", "leads", "lead", "new", "debate.", "said", "clinton", "poll.", "great", "big"_

In this topic, Donald Trump is tweeting about polling numbers, the debate and Hillary Clinton. These are likely tweets related to Donald Trump commenting and critizing what Hillary Clinton has said and how it has affected the polls.

In [13]:
# Model persistence: save(), load()
lsi.save('trump.lsi')
lsi = models.LsiModel.load('trump.lsi')

2019-02-16 20:37:05,410 : INFO : saving Projection object under trump.lsi.projection, separately None
2019-02-16 20:37:05,413 : INFO : saved trump.lsi.projection
2019-02-16 20:37:05,414 : INFO : saving LsiModel object under trump.lsi, separately None
2019-02-16 20:37:05,415 : INFO : not storing attribute projection
2019-02-16 20:37:05,417 : INFO : not storing attribute dispatcher
2019-02-16 20:37:05,418 : INFO : saved trump.lsi
2019-02-16 20:37:05,419 : INFO : loading LsiModel object from trump.lsi
2019-02-16 20:37:05,421 : INFO : loading id2word recursively from trump.lsi.id2word.* with mmap=None
2019-02-16 20:37:05,422 : INFO : setting ignored attribute projection to None
2019-02-16 20:37:05,423 : INFO : setting ignored attribute dispatcher to None
2019-02-16 20:37:05,424 : INFO : loaded trump.lsi
2019-02-16 20:37:05,425 : INFO : loading LsiModel object from trump.lsi.projection
2019-02-16 20:37:05,426 : INFO : loaded trump.lsi.projection


***
### Similarity

#### Step 1:

In [14]:
# Initializing the query structure: transform corpus to LSI space and index it
index = similarities.MatrixSimilarity(lsi[corpus])

2019-02-16 20:37:07,422 : WARNING : scanning corpus to determine the number of features (consider setting `num_features` explicitly)
2019-02-16 20:37:07,436 : INFO : creating matrix with 448 documents and 10 features


In [15]:
# Index persistence
index.save('trump.index')
index = similarities.MatrixSimilarity.load('trump.index')

2019-02-16 20:37:08,360 : INFO : saving MatrixSimilarity object under trump.index, separately None
2019-02-16 20:37:08,363 : INFO : saved trump.index
2019-02-16 20:37:08,365 : INFO : loading MatrixSimilarity object from trump.index
2019-02-16 20:37:08,367 : INFO : loaded trump.index


#### Step 2:

In [16]:
# Performing queries
doc = "Hillary Clinton."
vec_bow = bag.doc2bow(doc.lower().split())

# Convert the query to LSI space
vec_lsi = lsi[vec_bow]

# Perform a similarity query against the corpus
sims = index[vec_lsi]

# Ranking the tweets by their weights of similarity
sims = sorted(enumerate(sims), key=lambda item: -item[1])

# Printing the associated Tweets:
for i in range(10):
    print("Tweet Rank #{}:\tWeight: {}\nRaw text: {}\n".format(i+1, sims[i][1], document[sims[i][0]]))

Tweet Rank #1:	Weight: 0.9700218439102173
Raw text: Hillary Clinton is not a change agent, just the same old status quo! She is spending a fortune, I am spending very little. Close in polls!

Tweet Rank #2:	Weight: 0.959089994430542
Raw text: Don't believe the @FoxNews Polls, they are just another phony hit job on me. I will beat Hillary Clinton easily in the General Election.

Tweet Rank #3:	Weight: 0.8974069356918335
Raw text: .@USATODAY Poll and @QuinnipiacPoll say that I beat both Hillary and Bernie, and I havn't even started on them yet!

Tweet Rank #4:	Weight: 0.8799535036087036
Raw text: The Republican establishment, out of self preservation, is concerned w/ my high poll #'s. More concerned are Dems—I beat Hillary heads up!

Tweet Rank #5:	Weight: 0.8720756769180298
Raw text: Kasich only looks O.K. in polls against Hillary because nobody views him as a threat and therefore have placed ZERO negative ads against him

Tweet Rank #6:	Weight: 0.8531491756439209
Raw text: Ted Cruz is 

When we make a query for "Hillary Clinton" to retrieve the respective top tweets associated with her name, we find that the leading tweet with the greatest weight is a strong criticism of Hillary Clinton and her campaign spending. The remaining nine tweets are associated with the polls, darted with references to Donald Trump criticizing poll results not in his favor, him leading against the opposition as well as pitting himself likely to win.

In [17]:
# Performing queries
doc = "Ben Carson"
vec_bow = bag.doc2bow(doc.lower().split())

# Convert the query to LSI space
vec_lsi = lsi[vec_bow]

# Perform a similarity query against the corpus
sims = index[vec_lsi]

# Ranking the tweets by their weights of similarity
sims = sorted(enumerate(sims), key=lambda item: -item[1])

# Printing the associated Tweets:
for i in range(10):
    print("Tweet Rank #{}:\tWeight: {}\nRaw text: {}\n".format(i+1, sims[i][1], document[sims[i][0]]))

Tweet Rank #1:	Weight: 0.9365328550338745
Raw text: Just out: Boston Herald/Franklin Pierce Poll  N.H.  TRUMP 28 (up 10)  CARSON 16  BUSH 9  RUBIO 6  CRUZ 5  Press will say they are surging!

Tweet Rank #2:	Weight: 0.9363357424736023
Raw text: RT @DRUDGE_REPORT: REUTERS ROLLING:  TRUMP 39%, CRUZ 14.5%, BUSH 10.6%, CARSON 9.6%, RUBIO 6.7%... MORE... https://t.co/nRhtbzcqP9

Tweet Rank #3:	Weight: 0.9350287318229675
Raw text: Jeb Bush just got contact lenses and got rid of the glasses. He wants to look cool, but it's far too late. 1% in Nevada!

Tweet Rank #4:	Weight: 0.9347569346427917
Raw text: 2016 Republican Primary Morning Consult Poll was just released. TRUMP 32, CARSON 12, BUSH 11, FIORINA 6, RUBIO 5, CRUZ 5. Taken after debate

Tweet Rank #5:	Weight: 0.9299113154411316
Raw text: Tennessee GOP Poll
https://t.co/hfSxKaXC42
Trump 32.7%
Cruz 16.5%
Carson 6.6%
Rubio 5.3%
Christie 2.4%
Jeb 1.6%

Tweet Rank #6:	Weight: 0.9277924299240112
Raw text: CNN/ORC Poll results just out for Nevad

In [19]:
sims[1]

(98, 0.93633574)